c7n/resources/code.py

```diff
@@ -265,3 +265,71 @@ def process(self, resources):
                 self.manager.retry(client.delete_pipeline, name=r['name'])
             except client.exceptions.PipelineNotFoundException:
                 continue
+
+
+@resources.register('codedeploy-application')
+class CodeDeployApplication(QueryResourceManager):
+
+    class resource_type(TypeInfo):
+        service = 'codedeploy'
+        enum_spec = ('list_applications', 'applications', None)
+        batch_detail_spec = (
+            'batch_get_applications', 'applicationNames',
+            None, 'applicationsInfo', None)
+        id = 'applicationId'
+        name = 'applicationName'
+        date = 'createTime'
+        arn_type = ""
+        cfn_type = "AWS::CodeDeploy::Application"
+
+
+@CodeDeployApplication.action_registry.register('delete')
+class DeleteApplication(BaseAction):
+
+    schema = type_schema('delete')
+    permissions = ('codedeploy:DeleteApplication',)
+
+    def process(self, resources):
+        client = local_session(self.manager.session_factory).client('codedeploy')
+        for r in resources:
+            try:
+                self.manager.retry(client.delete_application, applicationName=r['applicationName'])
+            except client.exceptions.InvalidApplicationNameException:
+                continue
+
+
+@resources.register('codedeploy-deployment')
+class CodeDeployDeployment(QueryResourceManager):
+
+    class resource_type(TypeInfo):
+        service = 'codedeploy'
+        enum_spec = ('list_deployments', 'deployments', None)
+        batch_detail_spec = (
+            'batch_get_deployments', 'deploymentIds',
+            None, 'deploymentsInfo', None)
+        name = id = 'deploymentId'
+        # couldn't find a real cloudformation type
+        cfn_type = None
+        arn_type = ""
```

Comment:
deploymentgroup appears to be the closest arn type https://docs.aws.amazon.com/codedeploy/latest/userguide/auth-and-access-control-permissions-reference.html

Reply:
more specifically a deployment seems to always be qualified by both its application and deployment group. i'm curious what a deployment specifies here if it wasn't started with an explicit deployment group, if its a magic 'default' value for deployment group name/id. in some sense we're effectively skipping the optional parent, but perhaps thats okay (ie ancestry in the api appears to be app -> group -> deploy, but we're omitting group and going through all the data plane of deployments.

Reply:
it feels like some of this would enforcement would be better targeted to deployment group instead of the data plane on deployment to prevent mid deployment issues.

Reply:
Hello @kapilt , thanks for the feedback. I did some research on CodeDeploy and here's what I found:
1. To create a deployment group, it only requires the application name and deployment group name and my understanding is that `targetRevision` that we get back as part of [get_deployment_group](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/codedeploy.html#CodeDeploy.Client.get_deployment_group) api call inherits it from the actual deployment(which can change based on the actual deployment). I did not have to specify those details while creating a Deployment Group.
2. The [get_deployment_group](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/codedeploy.html#CodeDeploy.Client.get_deployment_group) api needs applicationName and DeploymentGroupName as part of required fields in the api which makes me think that it would have to be a child resource with Application as parent resulting in not having the luxury to have custodian monitor deployment groups in event based(CloudTrail) mode. Hence, I added a new deployment resource type that would monitor deployments in event based mode.
3.  I agree that we could add a DeploymentGroup Resource type later, but was wondering if we could get a cleaner api before we do that. Not sure if that's going to happen :)

Happy to hear your thoughts!